### PR TITLE
[cli] Add --relativize-paths-with option

### LINKED ISSUE #4026: [cli] Filenames printed as absolute paths in the report despite parameter `--short-names`
**Affects PMD Version:** 

PMD 6.47.0

**Description:**

The PMD CLI option `--short-names` is expected to print shortned filenames, however the filenames are printed in absolute format.
This feature was working fine 6.46.0

**Exception Stacktrace:**

```
AVERTISSEMENT: This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/pmd-6.47.0/pmd_userdocs_incremental_analysis.html
```

**Code Sample demonstrating the issue:**

```bash
pmd --rulesets ruleset.xml --file-list diff.txt --format xml --report-file pmd-report.xml --short-names
```

**Steps to reproduce:**

Please provide detailed steps for how we can reproduce the bug.

1.  Install pmd version 6.47.0 (latest)
2. Create a diff.txt file with list of files to be scanned
3. Create a ruleset file
4. Run analysis using the following command:
`pmd --rulesets ruleset.xml --file-list diff.txt --format xml --report-file pmd-report.xml --short-names`
5. Filenames in the pmd-report.xml are printed as absolute paths

**Running PMD through:** *[CLI]*

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

Add a new option to relativize paths (`--relativize-paths-with`, short form `-z`). The `--short-names` option does not change behavior but is deprecated.

Nothing changes if `-z` is not specified. In PMD 7 it might be smart to make the default either print paths as they are entered, or relativize based on current directory (`-z .`). Currently the default is to use absolute paths, which I think is not very intuitive, and which can be done with `-z /` anyway.

Question: currently I implemented `-z /` so that it renders absolute paths. Maybe it should actually render paths relative to `/`, which would be more regular, and we should have a separate option to render absolute paths?

## Related issues

- Fixes #4026

## Ready?

- [x] Release notes
- [x] Doc on website CLI reference
- [x] Test deprecation warning with `--short-names`
- [x] Test `-z /`
- [x] Decide whether to use a real `Path::relativize` or just remove a prefix (the prefix option is cleaner and also works better with zip files)
- [x] Prepare PMD 7 merge: #4363
- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/FileCollector.java [hunk 1]
```diff
@@ -310,21 +368,50 @@ public boolean addFileOrDirectory(Path file) throws IOException {
      * {@link #addFile(Path)} and such. The zip file is registered as
      * a resource to close at the end of analysis.
      */
-    public FileSystem addZipFile(Path zipFile) {
+    @Experimental
+    public void addZipFile(Path zipFile) throws IOException {
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/document/FileCollector.java:372]: This looks like an incompatible change: return type is removed now and a new checked exception is introduced.

Maybe we should keep the old method and deprecate it and create a new method...

File: pmd-core/src/test/java/net/sourceforge/pmd/cli/ZipFileTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    private final Path zipPath = Paths.get(ZIP_PATH);
+
+    @Test
+    public void testZipFile() {
+        PMDConfiguration conf = new PMDConfiguration();
+        conf.addInputPath(zipPath);
+        try (PmdAnalysis pmd = PmdAnalysis.create(conf)) {
+            List<TextFile> files = pmd.files().getCollectedFiles();
+            assertThat(files, hasSize(3));
+            assertThat(files.get(0).getDisplayName(), equalTo(ZIP_PATH + "!/otherSrc/somefile.dummy"));
```
Comment (adangel) [inline pmd-core/src/test/java/net/sourceforge/pmd/cli/ZipFileTest.java:33]: Since there is no `relativizeRoot` configured, I would expect here an absolute path for the zip file.

It seems, currently we implicitly rely on the current working directory for both input path and the created display name...

**Review by adangel — COMMENTED** [review]
Thanks! Looks promising :+1:

**Comment by adangel** [comment]
I've tested now, whether analyzing a zip file actually works. The results are:
* it worked in the past. The last working version is 6.43.0
* since 6.44.0 it is broken

Here's the output:

```shell
$ ~/PMD/binaries/pmd-bin-6.51.0/bin/run.sh pmd -d target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar -R rulesets/java/quickstart.xml -f text
Nov 24, 2022 9:53:36 AM net.sourceforge.pmd.util.log.internal.SimpleMessageReporter logImpl
SEVERE: Cannot open zip file target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar: Provider "zip" not found
Nov 24, 2022 9:53:37 AM net.sourceforge.pmd.PMD encourageToUseIncrementalAnalysis
WARNING: This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/pmd-6.51.0/pmd_userdocs_incremental_analysis.html
```

The stacktrace is:

```
FINE: java.nio.file.ProviderNotFoundException: Provider "zip" not found
	at java.base/java.nio.file.FileSystems.getFileSystem(FileSystems.java:238)
	at net.sourceforge.pmd.lang.document.FileCollector.addZipFile(FileCollector.java:319)
	at net.sourceforge.pmd.internal.util.FileCollectionUtil.addRoot(FileCollectionUtil.java:132)
	at net.sourceforge.pmd.internal.util.FileCollectionUtil.collectFiles(FileCollectionUtil.java:97)
	at net.sourceforge.pmd.internal.util.FileCollectionUtil.collectFiles(FileCollectionUtil.java:70)
	at net.sourceforge.pmd.PmdAnalysis.create(PmdAnalysis.java:113)
	at net.sourceforge.pmd.PMD.doPMD(PMD.java:247)
	at net.sourceforge.pmd.PMD.runPmd(PMD.java:505)
	at net.sourceforge.pmd.PMD.runPmd(PMD.java:472)
	at net.sourceforge.pmd.PMD.main(PMD.java:418)
```

According to https://docs.oracle.com/javase/7/docs/technotes/guides/io/fsp/zipfilesystemprovider.html the URL should use "jar" instead of "zip" to open the zip files.

When we analyzed zip file, we used by default the following paths in the report file (in PMD 6.43.0):

* current working directory: /home/andreas/PMD/source/pmd-it/MPMD-Smoke-Test/
  * params: `-d target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar` (relative path)
    * target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar:example/Sample.java
  * params: `-d target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar -shortnames` (relative path + shortnames)
    * target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar:example/Sample.java
  * params: `-d $PWD/target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar` (absolute path)
    * /home/andreas/PMD/source/pmd-it/MPMD-Smoke-Test/target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar:example/Sample.java
  * params: `-d $PWD/target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar -shortnames` (absolute path + shortnames)
    * maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar:example/Sample.java
* current working directory: /home/andreas
  * params: `-d PMD/source/pmd-it/MPMD-Smoke-Test/target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar` (relative path)
    * PMD/source/pmd-it/MPMD-Smoke-Test/target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar:example/Sample.java
  * params: `-d PMD/source/pmd-it/MPMD-Smoke-Test/target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar -shortnames` (relative path + shortnames)
    * PMD/source/pmd-it/MPMD-Smoke-Test/target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar:example/Sample.java
  * params: `-d $PWD/PMD/source/pmd-it/MPMD-Smoke-Test/target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar` (absolute path)
    * /home/andreas/PMD/source/pmd-it/MPMD-Smoke-Test/target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar:example/Sample.java
  * params: `-d $PWD/PMD/source/pmd-it/MPMD-Smoke-Test/target/maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar -shortnames` (absolute path + shortnames)
    * maven-pmd-plugin-it-Smoke-1.0-SNAPSHOT-sources.jar:example/Sample.java

In summary: shortnames only has an effect, if the path is provided absolutely with `-d`. And not the whole path is shortened - the name of the zip file remains.

Maybe I'll create an issue that we can't analyze zip files and add a test (a test was obviously missing for that feature otherwise we would have noticed...). This is separate from this PR, which focuses on shortnames/relativize paths.

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/b0551482688e43ada555e5e5a1bedd00/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/PMDTask.java [hunk 3]
```diff
@@ -264,4 +268,18 @@ public boolean isNoCache() {
     public void setNoCache(boolean noCache) {
         this.noCache = noCache;
     }
+
+    public void setRelativizePathsWith(String relativizePathsWith) {
```
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/ant/PMDTask.java:272]: TODO: Update https://pmd.github.io/latest/pmd_userdocs_tools_ant.html#parameters

File: pmd-core/src/main/java/net/sourceforge/pmd/PMDConfiguration.java [hunk 4]
```diff
@@ -940,4 +942,55 @@ public void setIgnoreIncrementalAnalysis(boolean noCache) {
...
+     * <p>If several paths are added, the shortest paths possible are
+     * built.
+     *
+     * @param path A path
+     *
+     * @throws IllegalArgumentException If the path points to a file, and not a directory
+     * @throws NullPointerException If the path is null
+     */
+    public void addRelativizeRoot(Path path) {
+        // TODO symlinks?
```
[the remark this answers is not in the document]
Comment (jsotuyod) [reply pmd-core/src/main/java/net/sourceforge/pmd/PMDConfiguration.java:961]: +1 to allowing symlinks and never resolve them. If the user inputs a path, we should output in terms of that path, not it's resolution.

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/PMDTask.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsotuyod) [reply pmd-core/src/main/java/net/sourceforge/pmd/ant/PMDTask.java:272]: yes, it should be `DirSet` to enforce it's a dir, and allow multiple entries

File: pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDParameters.java [hunk 5]
```diff
@@ -121,6 +125,17 @@ public class PMDParameters {
                    + "If this option is not specified, the report is rendered to standard output.")
     private String reportfile = null;
 
+    @Parameter(names = { RELATIVIZE_PATHS_WITH, "-z" },
+               variableArity = true,
+               description = "Path relative to which directories are rendered in the report."
+                             + "This option allows shortening directories in the report; "
+                             + "without it, paths are rendered as absolute paths. "
+                             + "The option can be repeated, in which case the shortest relative path will be used."
+                             + "This option replaces --short-names since PMD 6.54.0.",
```
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDParameters.java:134]: I'm unsure about having this here…

We are bound to forget removing it even after it stops making sense, the doc already states that, and at runtime we already have the suggested replacements being output as warning.

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/PMDTask.java [hunk 6]
```diff
@@ -25,7 +27,9 @@ public class PMDTask extends Task {
     private final List<FileSet> filesets = new ArrayList<>();
     private boolean failOnError;
     private boolean failOnRuleViolation;
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/ant/PMDTask.java:32]: I'm logging a warning now when this is used (see method validate())

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/PMDTask.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/ant/PMDTask.java]: I actually settled now to just use a path-like structure. Using a DirSet, it didn't look right to configure (e.g. setting a basedir to search in and then configuring a include pattern...)

File: pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDParameters.java (hunk 5 above)
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDParameters.java:134]: You mean the last sentence "This option replaces --short-names since PMD 6.54.0."?

Hm...yes, it would be now the only case, where we output it in the parameter help. We have it documented and we issue a deprecation warning, when "--short-names" is still used.

Yes, I guess, it's safe to remove this last sentence here.

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/FileCollector.java (hunk 1 above)
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/FileCollector.java]: I've added back the old method as deprecated and renamed the new one to `addZipFileWithContent`.

Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/FileCollector.java]: This new method returns a boolean, to behave like the other "add*" methods, which all return a boolean.

File: pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDParameters.java (hunk 5 above)
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDParameters.java]: I've removed this sentence now.

File: pmd-core/src/main/java/net/sourceforge/pmd/PMDConfiguration.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/PMDConfiguration.java]: I've added a comment here in the code stating, that there is no special handling for symlinks.

While adding tests, I noticed, that we didn't follow symlinks yet, when collecting source files, so I enabled this (see the change in `FileCollector::addDirectory`)

File: pmd-core/src/test/java/net/sourceforge/pmd/cli/ZipFileTest.java (hunk 2 above)
Comment (adangel) [reply pmd-core/src/test/java/net/sourceforge/pmd/cli/ZipFileTest.java]: Actually reviewed the tests and kept the implementation as is.

Updated the documentation: If no relative-root-path is given, the reported path is the same like given as in `--dir`. I think, that's the least surprising behavior.

Added then back the sentence to use "/" to get absolute paths (we already had tests for this in ZipFileTest, added more tests in CoreCliTest)

**Comment by adangel** [comment]
Im ticking off

> Decide whether to use a real Path::relativize or just remove a prefix (the prefix option is cleaner and also works better with zip files)

The current implementation uses indeed `Path::relativize` - and it works. So I see no reason, why to change that now. The tests cover also zip files (although not relativizing within zip files, but I don't want to go that deep).

I'll prepare the merge to pmd/7.0.x next...

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/FileCollector.java [hunk 7]
```diff
@@ -277,7 +335,7 @@ public boolean addDirectory(Path dir) throws IOException {
             reporter.error("Not a directory {0}", dir);
             return false;
         }
-        Files.walkFileTree(dir, new SimpleFileVisitor<Path>() {
+        Files.walkFileTree(dir, EnumSet.of(FileVisitOption.FOLLOW_LINKS), Integer.MAX_VALUE, new SimpleFileVisitor<Path>() {
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/document/FileCollector.java:338]: That's the change, that will now follow symlinks, if someone provides a symlink as `--dir`... should we do this change?

Comment (jsotuyod) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/FileCollector.java:338]: I think so, yes

**Comment by adangel** [comment]
@jsotuyod I'm taking care of merging master->pmd/7.0.x...
